design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+6. We should be able to drop out unused layer types with a compile flag.
+
+###  <a name="eval">Evaluation Metrics</a>
+
+1. Developers should be able to continue using the OpenGL renderer with this refactor.
+2. The resulting binary from the modularized refactor should not increase more than 5% when compiled.
+3. The modularization refactor should be visually identical to the current implementation when using OpenGL. All existing render tests should pass, and developers that continue to use this library with an OpenGL renderer should not experience any visual degradation such as artifacts.
+4. After refactor, the startup time for applications built with this library should not increase. This will be measured by the startup time of [this](https://github.com/maplibre/maplibre-gl-native/blob/main/platform/ios/platform/ios/iosapp%20UITests/iosapp_UITests.swift) iOS test app.
+5. When implemented with a given style and set of tilesets, the compiled library should not decrease the rendering frame rate.
+6. This refactor will not introduce data loading or other bottlenecks that will negatively impact rendering performance.
+7. Developers currently using this library will not see any breaking API changes, and will be able to seamlessly upgrade versions.
+
+## Proposed Change
+
+### Shader(Program) Representation
+
+_Addresses core functionality [#1](#core) and [#2](#core)._
+
+We need an external representation for programs (shaders) so they can be added or replaced by developers.  
+
+Internally to the toolkit shaders are called Programs, but that's far too confusing at this level.  Are we referring to a GPU program or a program the user writes?  Best to stick with the term shader.
+
+The shaders need a representation visible outside the toolkit and we need to change the way they’re handled inside the toolkit.  We need to be able to replace them, add brand new ones, and allow developers to control their shaders.
+
+#### The way it is now
+
+Individual shaders are represented by an object class with massive amounts of template logic and a minimum of in-line comments.  They’re opaque from the outside and only controllable through vector tile data and styles. 
+
+The [render_raster_layer](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/layers/render_raster_layer.cpp), as an example, asks for the instantiation of (eventually) RasterProgram.  Rather than having the source for the program, that then pokes into a compressed chunk of memory that contains the source, which is then uncompressed and fed into OpenGL for compilation.
+
+All of this makes sense in context, but is completely inaccessible outside the toolkit and rather incomprehensible when making changes.
+
+#### Required changes
+To let developers control their own shaders we’ll need to:
+- Allow shaders to be instantiated with shader source (one option).
```

Comment:
Which source format do we have in mind?

I guess just by adding a shader compiler to the binary it will already increase by 5%.
So maybe we should state above that the size should not increase above 5% with the same features like before (without dynamic shader compiliation)

Reply:
The current implementation is a bit odd.  There's a compressed bundle of text which is decompressed and then run through the OpenGL shader compiler.  That compiler mechanism is always going to be present on the device, so no cost to the toolkit there.  It is a bit of time lost, but not very much and pre-compiling shaders in OpenGL is somewhat dicey.

Later on we propose allowing the current text blob and a new version where the shader source is readable in the project.  This will still be compressed by the mobile devices, so I'm not sure it comes out all that different in size.  In any case we'll allow either option by compiler flag, so we can fall back on the old one.